File: backend/app/routers/writing_templates.py
from typing import Optional
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session
from app.db.session import get_db
from app.services.writing_template_service import writing_template_library
from app.models.user import User
from app.core.security import get_current_user_required

router = APIRouter(prefix="/writing-templates", tags=["writing-templates"])


@router.get("/genres")
def list_genre_templates():
    """List all available genre templates."""
    genres = writing_template_library.get_all_genres()
    templates = []
    for genre in genres:
        template = writing_template_library.get_genre_template(genre)
        templates.append({
            "genre": genre,
            "name": template["name"],
            "description": template["description"],
            "chapter_count": len(template["structure"]),
        })
    return {"templates": templates}


@router.get("/genres/{genre}")
def get_genre_template(genre: str):
    """Get detailed template for a specific genre."""
    template = writing_template_library.get_genre_template(genre)
    if not template:
        return {"error": f"Template not found for genre: {genre}"}
    return template


@router.get("/structures")
def list_story_structures():
    """List all available story structures."""
    structures = writing_template_library.get_all_structures()
    result = []
    for structure in structures:
        template = writing_template_library.get_structure_template(structure)
        result.append({
            "structure": structure,
            "name": template["name"],
            "description": template["description"],
        })
    return {"structures": result}


@router.get("/structures/{structure}")
def get_story_structure(structure: str):
    """Get detailed story structure template."""
    template = writing_template_library.get_structure_template(structure)
    if not template:
        return {"error": f"Structure not found: {structure}"}
    return template


@router.post("/generate-outline")
def generate_outline_from_template(
    genre: str,
    custom_title: Optional[str] = None,
    num_chapters: Optional[int] = 10,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user_required),
):
    """Generate a chapter outline from a genre template."""
    result = writing_template_library.generate_outline_from_template(
        genre=genre,
        custom_title=custom_title,
        num_chapters=num_chapters,
    )
    return result

File: frontend/src/api/writingTemplates.ts
import client from './client';

export interface GenreTemplate {
  genre: string;
  name: string;
  description: string;
  structure: Array<{
    chapter: number;
    title: string;
    description: string;
  }>;
  themes: string[];
  tropes: string[];
}

export interface StoryStructure {
  name: string;
  description: string;
  acts?: Array<{
    act: number;
    name: string;
    chapters: string;
    description: string;
  }>;
  stages?: Array<{
    stage: string;
    description: string;
  }>;
}

export interface GeneratedOutline {
  genre: string;
  template_name: string;
  chapters: Array<{
    chapter_number: number;
    title: string;
    description: string;
  }>;
  themes: string[];
  tropes: string[];
}

export const writingTemplatesApi = {
  // List all genre templates
  listGenres: () =>
    client.get<{
      templates: Array<{
        genre: string;
        name: string;
        description: string;
        chapter_count: number;
      }>;
    }>('/writing-templates/genres'),

  // Get detailed genre template
  getGenreTemplate: (genre: string) =>
    client.get<GenreTemplate>(`/writing-templates/genres/${genre}`),

  // List story structures
  listStructures: () =>
    client.get<{
      structures: Array<{
        structure: string;
        name: string;
        description: string;
      }>;
    }>('/writing-templates/structures'),

  // Get detailed story structure
  getStoryStructure: (structure: string) =>
    client.get<StoryStructure>(`/writing-templates/structures/${structure}`),

  // Generate outline from template
  generateOutline: (data: {
    genre: string;
    custom_title?: string;
    num_chapters?: number;
  }) =>
    client.post<GeneratedOutline>('/writing-templates/generate-outline', data, {
      params: {
        genre: data.genre,
        custom_title: data.custom_title,
        num_chapters: data.num_chapters,
      },
    }),
};
